File: chatfeature/src/main/java/com/ekoapp/sample/chatfeature/messagelist/MessageListAdapter.java
package com.ekoapp.sample.chatfeature.messagelist;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;

import com.bumptech.glide.Glide;
import com.ekoapp.ekosdk.EkoMessage;
import com.ekoapp.ekosdk.EkoObjects;
import com.ekoapp.ekosdk.EkoUser;
import com.ekoapp.ekosdk.adapter.EkoMessageAdapter;
import com.ekoapp.ekosdk.messaging.data.DataType;
import com.ekoapp.ekosdk.messaging.data.FileData;
import com.ekoapp.ekosdk.messaging.data.ImageData;
import com.ekoapp.ekosdk.messaging.data.TextData;
import com.ekoapp.sample.chatfeature.BaseViewHolder;
import com.ekoapp.sample.chatfeature.R;
import com.google.common.base.Joiner;

import org.joda.time.DateTime;
import org.joda.time.format.DateTimeFormat;

import java.util.List;

import butterknife.BindView;
import butterknife.BindViews;
import butterknife.Setter;
import butterknife.ViewCollections;
import io.reactivex.BackpressureStrategy;
import io.reactivex.Flowable;
import io.reactivex.subjects.PublishSubject;

public class MessageListAdapter extends EkoMessageAdapter<MessageListAdapter.MessageViewHolder> {

    private final PublishSubject<EkoMessage> onLongClickSubject = PublishSubject.create();
    private final PublishSubject<EkoMessage> onClickSubject = PublishSubject.create();


    @NonNull
    @Override
    public MessageViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.item_message, parent, false);
        return new MessageViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull MessageViewHolder holder, int position) {
        EkoMessage m = getItem(position);

        Setter<View, Integer> visibility = (view, value, index) -> view.setVisibility(value);

        if (EkoObjects.isProxy(m)) {
            renderLoadingItem(holder, visibility, position);
        } else if (m.isDeleted()) {
           renderDeletedMessage(holder, visibility);
        } else {
            ViewCollections.set(holder.optionalViews, visibility, View.VISIBLE);
            String type = m.getType();
            EkoUser sender = m.getUser();
            DateTime created = m.getCreatedAt();

            holder.messageIdTextview.setText(String.format("mid: %s %s:%s\nsegment: %s",
                    m.getMessageId(),
                    m.isFlaggedByMe() ? "\uD83C\uDFC1" : "\uD83C\uDFF3️",
                    m.getFlagCount(),
                    m.getChannelSegment()));

            holder.senderTextview.setText(String.format("uid: %s %s: %s\ndisplay name: %s",
                    sender != null ? sender.getUserId() : "",
                    sender != null && sender.isFlaggedByMe() ? "\uD83C\uDFC1" : "\uD83C\uDFF3️",
                    sender != null ? sender.getFlagCount() : 0,
                    sender != null ? sender.getDisplayName() : ""));

            holder.commentTextview.setText(String.format("comment count: %s", m.getChildrenNumber()));
            holder.tagsTextview.setText(String.format("tags️: %s", Joiner.on(", ").join(m.getTags())));

            if (DataType.TEXT == DataType.from(m.getType())) {
                holder.dataTextview.setText(String.format("text: %s", m.getData(TextData.class).getText()));
                Glide.with(holder.dataImageview.getContext()).clear(holder.dataImageview);

            } else if (DataType.IMAGE == DataType.from(m.getType())) {
                String url = m.getData(ImageData.class).getUrl();
                holder.dataTextview.setText(String.format("data type: %s,\nurl: %s,\ndata: ", type, url).concat(m.getData().toString()));
                Glide.with(holder.dataImageview.getContext()).load(url).into(holder.dataImageview);

            } else if (DataType.FILE == DataType.from(m.getType())) {
                String url = m.getData(FileData.class).getUrl();
                holder.dataTextview.setText(String.format("data type: %s,\nurl: %s,\ndata: ", type, url).concat(m.getData().toString()));
                Glide.with(holder.dataImageview.getContext()).clear(holder.dataImageview);
            } else {
                holder.dataTextview.setText(String.format("data type: %s,\ndata: ", type).concat(m.getData().toString()));
                Glide.with(holder.dataImageview.getContext()).clear(holder.dataImageview);
            }

            renderReaction(holder, m);

            holder.syncStateTextview.setText(m.getSyncState().name());
            holder.timeTextview.setText(created.toString(DateTimeFormat.longDateTime()));
        }

        holder.itemView.setOnLongClickListener(v -> {
            onLongClickSubject.onNext(m);
            return true;
        });

        holder.itemView.setOnClickListener(v -> {
            onClickSubject.onNext(m);
        });
    }

    private void renderLoadingItem(@NonNull MessageViewHolder holder, Setter<View, Integer> visibility, int position) {
        ViewCollections.set(holder.optionalViews, visibility, View.GONE);
        holder.messageIdTextview.setText(String.format("loading adapter position: %s", position));
        Glide.with(holder.dataImageview.getContext()).clear(holder.dataImageview);
    }

    private void renderDeletedMessage(@NonNull MessageViewHolder holder, Setter<View, Integer> visibility) {
        ViewCollections.set(holder.optionalViews, visibility, View.GONE);
        holder.messageIdTextview.setText(String.format("Deleted"));
        holder.dataTextview.setText("");
        Glide.with(holder.dataImageview.getContext()).clear(holder.dataImageview);
    }

    private void renderReaction(@NonNull MessageViewHolder holder, @NonNull EkoMessage m) {
        String reactions = "";
        for(String key: m.getReactions().keySet()) {
            int reactionCount = m.getReactions().getCount(key);
            reactions += String.format("\n%s : %s", key, reactionCount);
        }

        List<String> myReactions = m.getMyReactions();
        String myReactionsString = Joiner.on(" ").join(myReactions);

        holder.reactionTextview.setText(String.format("reaction count: %s %s \nmy reactions: %s",
                m.getReactionCount(),
                reactions,
                myReactionsString));
    }


    Flowable<EkoMessage> getOnLongClickFlowable() {
        return onLongClickSubject.toFlowable(BackpressureStrategy.BUFFER);
    }

    Flowable<EkoMessage> getOnClickFlowable() {
        return onClickSubject.toFlowable(BackpressureStrategy.BUFFER);
    }


    static class MessageViewHolder extends BaseViewHolder {

        @BindViews({
                R.id.sender_textview,
                R.id.data_textview,
                R.id.sync_state_textview,
                R.id.time_textview
        })
        List<View> optionalViews;

        @BindView(R.id.message_textview)
        TextView messageIdTextview;
        @BindView(R.id.sender_textview)
        TextView senderTextview;
        @BindView(R.id.data_textview)
        TextView dataTextview;
        @BindView(R.id.data_imageview)
        ImageView dataImageview;
        @BindView(R.id.reaction_textview)
        TextView reactionTextview;
        @BindView(R.id.comment_count_textview)
        TextView commentTextview;
        @BindView(R.id.tags_textview)
        TextView tagsTextview;
        @BindView(R.id.sync_state_textview)
        TextView syncStateTextview;
        @BindView(R.id.time_textview)
        TextView timeTextview;

        MessageViewHolder(View itemView) {
            super(itemView);
        }
    }

}

File: chat_feature/src/main/java/com/ekoapp/sample/chatfeature/messagelist/MessageListAdapter.kt
package com.ekoapp.sample.chatfeature.messagelist

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import com.bumptech.glide.Glide
import com.ekoapp.ekosdk.EkoMessage
import com.ekoapp.ekosdk.EkoObjects
import com.ekoapp.ekosdk.adapter.EkoMessageAdapter
import com.ekoapp.ekosdk.messaging.data.DataType
import com.ekoapp.ekosdk.messaging.data.FileData
import com.ekoapp.ekosdk.messaging.data.ImageData
import com.ekoapp.ekosdk.messaging.data.TextData
import com.ekoapp.sample.chatfeature.R
import com.ekoapp.sample.chatfeature.messagelist.MessageListAdapter.MessageViewHolder
import com.google.common.base.Joiner
import io.reactivex.BackpressureStrategy
import io.reactivex.Flowable
import io.reactivex.subjects.PublishSubject
import kotlinx.android.synthetic.main.item_message.view.*
import org.joda.time.format.DateTimeFormat

class MessageListAdapter : EkoMessageAdapter<MessageViewHolder>() {
    private val onLongClickSubject = PublishSubject.create<EkoMessage>()
    private val onClickSubject = PublishSubject.create<EkoMessage>()
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): MessageViewHolder {
        val view: View = LayoutInflater.from(parent.context).inflate(R.layout.item_message, parent, false)
        return MessageViewHolder(view)
    }

    override fun onBindViewHolder(holder: MessageViewHolder, position: Int) {
        val m = getItem(position)
        if (EkoObjects.isProxy(m)) {
            renderLoadingItem(holder, position)
        } else if (m!!.isDeleted) {
            renderDeletedMessage(holder)
        } else {
            setViewsVisibility(holder.optionalViews, View.VISIBLE)
            val type = m.type
            val sender = m.user
            val created = m.createdAt
            holder.itemView.message_textview.text = String.format("mid: %s %s:%s\nsegment: %s",
                    m.messageId,
                    if (m.isFlaggedByMe) "\uD83C\uDFC1" else "\uD83C\uDFF3️",
                    m.flagCount,
                    m.channelSegment)
            holder.itemView.sender_textview.text = String.format("uid: %s %s: %s\ndisplay name: %s",
                    sender?.userId ?: "",
                    if (sender != null && sender.isFlaggedByMe) "\uD83C\uDFC1" else "\uD83C\uDFF3️",
                    sender?.flagCount ?: 0,
                    if (sender != null) sender.displayName else "")
            holder.itemView.comment_count_textview.text = String.format("comment count: %s", m.childrenNumber)
            holder.itemView.tags_textview.text = String.format("tags️: %s", Joiner.on(", ").join(m.tags))
            if (DataType.TEXT == DataType.from(m.type)) {
                holder.itemView.data_textview.text = String.format("text: %s", m.getData(TextData::class.java).text)
                Glide.with(holder.itemView.data_imageview!!.context).clear(holder.itemView.data_imageview)
            } else if (DataType.IMAGE == DataType.from(m.type)) {
                val url = m.getData(ImageData::class.java).url
                holder.itemView.data_textview.text = String.format("data type: %s,\nurl: %s,\ndata: ", type, url) + m.data.toString()
                Glide.with(holder.itemView.data_imageview.context).load(url).into(holder.itemView.data_imageview)
            } else if (DataType.FILE == DataType.from(m.type)) {
                val url = m.getData(FileData::class.java).url
                holder.itemView.data_textview.text = String.format("data type: %s,\nurl: %s,\ndata: ", type, url) + m.data.toString()
                Glide.with(holder.itemView.data_imageview.context).clear(holder.itemView.data_imageview)
            } else {
                holder.itemView.data_textview.text = String.format("data type: %s,\ndata: ", type) + m.data.toString()
                Glide.with(holder.itemView.data_imageview.context).clear(holder.itemView.data_imageview)
            }
            renderReaction(holder, m)
            holder.itemView.sync_state_textview.text = m.syncState.name
            holder.itemView.time_textview.text = created.toString(DateTimeFormat.longDateTime())
        }
        holder.itemView.setOnLongClickListener { v: View? ->
            onLongClickSubject.onNext(m!!)
            true
        }
        holder.itemView.setOnClickListener { v: View? -> onClickSubject.onNext(m!!) }
    }

    private fun renderLoadingItem(holder: MessageViewHolder, position: Int) {
        setViewsVisibility(holder.optionalViews, View.GONE)
        holder.itemView.message_textview.text = String.format("loading adapter position: %s", position)
        Glide.with(holder.itemView.data_imageview.context).clear(holder.itemView.data_imageview)
    }

    private fun renderDeletedMessage(holder: MessageViewHolder) {
        setViewsVisibility(holder.optionalViews, View.GONE)
        holder.itemView.message_textview.text = String.format("Deleted")
        holder.itemView.data_textview.text = ""
        Glide.with(holder.itemView.data_imageview.context).clear(holder.itemView.data_imageview)
    }

    private fun renderReaction(holder: MessageViewHolder, m: EkoMessage) {
        var reactions = ""
        for (key in m.reactions.keys) {
            val reactionCount = m.reactions.getCount(key)
            reactions += String.format("\n%s : %s", key, reactionCount)
        }
        val myReactions = m.myReactions
        val myReactionsString = Joiner.on(" ").join(myReactions)
        holder.itemView.reaction_textview.text = String.format("reaction count: %s %s \nmy reactions: %s",
                m.reactionCount,
                reactions,
                myReactionsString)
    }

    val onLongClickFlowable: Flowable<EkoMessage>
        get() = onLongClickSubject.toFlowable(BackpressureStrategy.BUFFER)

    val onClickFlowable: Flowable<EkoMessage>
        get() = onClickSubject.toFlowable(BackpressureStrategy.BUFFER)

    class MessageViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        var optionalViews: List<View> = listOf(itemView.sender_textview,
                itemView.data_textview,
                itemView.sync_state_textview,
                itemView.time_textview)
    }

    private fun setViewsVisibility(views: List<View>, visibility: Int) {
        views.forEach {
            it.visibility = visibility
        }
    }

}
